src/core/quantum_gate.jl

```diff
@@ -147,6 +155,145 @@ function apply_gate!(state::Ket, gate::Gate)
     Snowflake.apply_gate_without_ket_size_check!(state, gate, Int(qubit_count))
 end
 
+function apply_gate!(state::Ket, gate::AbstractGate)
+    qubit_count = log2(length(state))
+    if mod(qubit_count, 1) != 0
+        throw(DomainError(qubit_count,
+            "Ket does not correspond to an integer number of qubits"))
+    end
+    
+    if any(t -> t>qubit_count ,gate.target_list)
+        throw(DomainError(gate.target_list,
+            "not enough qubits in the Ket for the targets in gate"))
+    end
+
+    connected_qubits=get_connected_qubits(gate)
+
+    type_in_ket=eltype(state.data)
+
+    operator=get_operator(gate,type_in_ket)
+
+    apply_operator!(state,operator,connected_qubits,Int(qubit_count),Val(length(connected_qubits)))
```

Comment:
The operator already has the size fixed at compile time. Is it necessary to pass the number of connected qubits as a parameter?